compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtAndMacroOperandsInSameScope.ts

```diff
@@ -42,250 +67,210 @@ import {Iterable_some} from '../Utils/utils';
-          macroValues.has(operand.identifier),
-        )
-      ) {
-        const macroOperands: Array<Identifier> = [];
-        for (const operand of eachInstructionValueOperand(value)) {
-          if (macroValues.has(operand.identifier)) {
-            macroOperands.push(operand.identifier);
+        case 'PropertyLoad': {
+          if (typeof value.property === 'string') {
+            const macroDefinition = macroTags.get(value.object.identifier.id);
+            if (macroDefinition != null) {
+              const propertyDefinition =
+                macroDefinition.properties != null
+                  ? (macroDefinition.properties.get(value.property) ??
+                    macroDefinition.properties.get('*'))
+                  : null;
+              const propertyMacro = propertyDefinition ?? macroDefinition;
+              macroTags.set(lvalue.identifier.id, propertyMacro);
+            }
           }
+          break;
         }
-        macroValues.set(lvalue.identifier, macroOperands);
       }
     }
   }
+  return macroTags;
 }
 
-function mergeScopes(
-  root: Identifier,
-  scope: ReactiveScope,
-  macroValues: Map<Identifier, Array<Identifier>>,
-  macroTagsCalls: Set<IdentifierId>,
-): void {
-  const operands = macroValues.get(root);
-  if (operands == null) {
-    return;
-  }
-  for (const operand of operands) {
-    operand.scope = scope;
-    expandFbtScopeRange(scope.range, operand.mutableRange);
-    macroTagsCalls.add(operand.id);
-    mergeScopes(operand, scope, macroValues, macroTagsCalls);
-  }
-}
-
-function matchesExactTag(s: string, tags: Set<Macro>): boolean {
-  return Array.from(tags).some(macro =>
-    typeof macro === 'string'
-      ? s === macro
-      : macro[1].length === 0 && macro[0] === s,
-  );
-}
-
-function matchTagRoot(
-  s: string,
-  tags: Set<Macro>,
-): Array<Array<MacroMethod>> | null {
-  const methods: Array<Array<MacroMethod>> = [];
-  for (const macro of tags) {
-    if (typeof macro === 'string') {
-      continue;
+function mergeMacroArguments(
+  fn: HIRFunction,
+  macroTags: Map<IdentifierId, MacroDefinition>,
+  macroKinds: Map<Macro, MacroDefinition>,
+): Set<IdentifierId> {
+  const macroValues = new Set<IdentifierId>(macroTags.keys());
+  for (const block of Array.from(fn.body.blocks.values()).reverse()) {
+    for (let i = block.instructions.length - 1; i >= 0; i--) {
+      const instr = block.instructions[i]!;
+      const {lvalue, value} = instr;
+      switch (value.kind) {
+        case 'DeclareContext':
+        case 'DeclareLocal':
+        case 'Destructure':
+        case 'LoadContext':
+        case 'LoadLocal':
+        case 'PostfixUpdate':
+        case 'PrefixUpdate':
+        case 'StoreContext':
+        case 'StoreLocal': {
+          // Instructions that never need to be merged
+          break;
+        }
+        case 'CallExpression':
+        case 'MethodCall': {
+          const scope = lvalue.identifier.scope;
+          if (scope == null) {
```

Comment:
In following case,
- we're processing a transitive macro (e.g. imagine a function that returned a boolean but expected its arguments to be ast-static)
- lvalue is unscoped due to type inference knowing it is a primitive type
- the function call arguments are scoped

```
// ModuleDefinition: `macro` is non-mutating and returns a primitive
const isEnabled: boolean = macro('MyFeatureFlag', { staticKey: "staticValue" });
```
we may need to prune the scope of `{ staticKey: "staticValue" }`